<a href="https://colab.research.google.com/github/DarthSaurus/Bedu/blob/Analisis-de-Datos-con-Python/Sesion-01/Reto-03/desviacion_estandar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Reto 3: Desviación estándar

### 1. Objetivos:
    - Utilizar la desviación estándar para realizar un análisis de dispersión de nuestros datos

---
    
### 2. Desarrollo:

#### a) Desviación estándar y distribución de los datos

Como ya vimos, la desviación estándar es la medida que nos da la "desviación típica" (o esperada) de nuestros datos a comparación del promedio. Eso quiere decir que normalmente vamos a esperar que una gran parte de nuestros datos se encuentren a 1 desviación estándar de distancia del promedio. Entre más nos alejamos, menos muestras deberíamos de encontrar.

Vamos a comprobar esto usando nuestro dataset de meteoritos que orbitan cerca de la Tierra. Tu Reto consiste en los siguientes pasos:

1. Crea un DataFrame con el dataset 'near_earth_objects-jan_feb_1995-clean.csv'.
2. Obtén la cantidad total de muestras en tu DataFrame.
3. Obtén la desviación estándar de la columna 'estimated_diameter.meters.estimated_diameter_max'. Los siguientes pasos realízalos todos utilizando esta columna.
4. Obtén el porcentaje de muestras que están a una distancia de 1 desviación estándar del promedio.
5. Obtén el porcentaje de muestras que están a una distancia de 2 desviaciones estándares del promedio.
6. Obtén el porcentaje de muestras que están a una distancia de 3 desviaciones estándares del promedio.
7. Compara los porcentajes obtenidos y comenta con tus compañeros y la experta tus hallazgos. ¿Qué significa esto? ¿La definición de desviación estándar tiene sentido? ¿Qué puedo inferir acerca de la dispersión de mis datos a partir de los valores obtenidos?

> Nota: Para obtener los porcentajes de los subconjuntos primero necesitas filtrar el DataFrame original para que sólo permanezcan las muestras que cumplan con los requisitos.

> Nota: Este Reto está diseñado para tener una dificultad media. No te frustres si al principio parece demasiado difícil. Comienza poco a poco, resolviendo el problema en pedazos pequeños, y si no tienes la menor idea de cómo proceder recuerda que la experta está ahí para ayudarte.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Crearé un DataFrame con el dataset 'near_earth_objects-jan_feb_1995-clean.csv'.
Observaré las primeras entradas para familiarizarme con la estructura de los datos.

In [3]:
### Tu código va aquí
import pandas as pd

# Load the dataset
df_meteorites = pd.read_csv('/content/drive/MyDrive/Remoto Datasets/Remoto near_earth_objects-jan_feb_1995-clean.csv')

# Display the first few rows of the dataframe
df_meteorites.head()


Unnamed: 0       id                name  is_potentially_hazardous_asteroid  \
0           0  2154652  154652 (2004 EP20)                              False   
1           1  3153509           (2003 HM)                               True   
2           2  3516633           (2010 HA)                              False   
3           3  3837644          (2019 AY3)                              False   
4           4  3843493           (2019 PY)                              False   

   estimated_diameter.meters.estimated_diameter_min  \
0                                        483.676488   
1                                         96.506147   
2                                         44.111820   
3                                         46.190746   
4                                         22.108281   

   estimated_diameter.meters.estimated_diameter_max close_approach_date  \
0                                       1081.533507          1995-01-07   
1                                        215.794305          1995-01-07   
2                                         98.637028          1995-01-07   
3                                        103.285648          1995-01-07   
4                                         49.435619          1995-01-07   

   epoch_date_close_approach orbiting_body  \
0               789467580000         Earth   
1               789491340000         Earth   
2               789446820000         Earth   
3               789513900000         Earth   
4               789446700000         Earth   

   relative_velocity.kilometers_per_second  \
0                                16.142864   
1                                12.351044   
2                                 6.220435   
3                                22.478615   
4                                 4.998691   

   relative_velocity.kilometers_per_hour  
0                           58114.308667  
1                           44463.757734  
2                           22393.567277  
3                           80923.015021  
4                           17995.288355

Obtener la cantidad total de muestras en el DataFrame.
Calcular la desviación estándar de la columna 'estimated_diameter.meters.estimated_diameter_max'.

In [4]:
# 2. Get the total number of samples in the dataframe
total_samples = len(df_meteorites)

# 3. Calculate the standard deviation of the column 'estimated_diameter.meters.estimated_diameter_max'
std_dev = df_meteorites['estimated_diameter.meters.estimated_diameter_max'].std()

total_samples, std_dev


(333, 614.6915918552232)

Tenemos un total de 333 muestras en el DataFrame. La desviación estándar de la columna 'estimated_diameter.meters.estimated_diameter_max' es aproximadamente
614.69
614.69 metros.

<details><summary>Solución</summary>

```python
import pandas as pd

df = pd.read_csv('../../Datasets/near_earth_objects-jan_feb_1995-clean.csv', index_col=0)
diameter_column = 'estimated_diameter.meters.estimated_diameter_max'
total_count = df.shape[0]
mean = df[diameter_column].mean()
std = df[diameter_column].std()

within_one_std_filter_bottom = df[diameter_column] >= (mean - std)
within_one_std_filter_top = df[diameter_column] <= (mean + std)
within_one_std = df[within_one_std_filter_bottom & within_one_std_filter_top]
percentage_of_data_within_one_std = within_one_std.shape[0] * 100 / total_count

within_two_std_filter_bottom = df[diameter_column] >= (mean - 2 * std)
within_two_std_filter_top = df[diameter_column] <= (mean + 2 * std)
within_two_std = df[within_two_std_filter_bottom & within_two_std_filter_top]
percentage_of_data_within_two_std = within_two_std.shape[0] * 100 / total_count

within_three_std_filter_bottom = df[diameter_column] >= (mean - 3 * std)
within_three_std_filter_top = df[diameter_column] <= (mean + 3 * std)
within_three_std = df[within_three_std_filter_bottom & within_three_std_filter_top]
percentage_of_data_within_three_std = within_three_std.shape[0] * 100 / total_count
```
    
</details>

A continuación, calcularemos los porcentajes de muestras que se encuentran a una distancia de 1, 2 y 3 desviaciones estándar del promedio:

Obtener el porcentaje de muestras que están a una distancia de 1 desviación estándar del promedio.

Obtener el porcentaje de muestras que están a una distancia de 2 desviaciones estándares del promedio.

Obtener el porcentaje de muestras que están a una distancia de 3 desviaciones estándares del promedio.

In [5]:
# 4. Percentage of samples within 1 standard deviation from the mean
mean_value = df_meteorites['estimated_diameter.meters.estimated_diameter_max'].mean()
samples_within_1std = df_meteorites[(df_meteorites['estimated_diameter.meters.estimated_diameter_max'] >= mean_value - std_dev) &
                                   (df_meteorites['estimated_diameter.meters.estimated_diameter_max'] <= mean_value + std_dev)]
percentage_1std = (len(samples_within_1std) / total_samples) * 100

# 5. Percentage of samples within 2 standard deviations from the mean
samples_within_2std = df_meteorites[(df_meteorites['estimated_diameter.meters.estimated_diameter_max'] >= mean_value - 2*std_dev) &
                                   (df_meteorites['estimated_diameter.meters.estimated_diameter_max'] <= mean_value + 2*std_dev)]
percentage_2std = (len(samples_within_2std) / total_samples) * 100

# 6. Percentage of samples within 3 standard deviations from the mean
samples_within_3std = df_meteorites[(df_meteorites['estimated_diameter.meters.estimated_diameter_max'] >= mean_value - 3*std_dev) &
                                   (df_meteorites['estimated_diameter.meters.estimated_diameter_max'] <= mean_value + 3*std_dev)]
percentage_3std = (len(samples_within_3std) / total_samples) * 100

percentage_1std, percentage_2std, percentage_3std


(90.39039039039038, 96.3963963963964, 97.8978978978979)

Aquí están los porcentajes de muestras que se encuentran dentro de 1, 2 y 3 desviaciones estándar del promedio:

Aproximadamente el
90.39
%
90.39% de las muestras están dentro de 1 desviación estándar del promedio.
Aproximadamente el
96.40
%
96.40% de las muestras están dentro de 2 desviaciones estándar del promedio.
Aproximadamente el
97.90
%
97.90% de las muestras están dentro de 3 desviaciones estándar del promedio.
Estos resultados están en línea con la regla empírica (o regla del 68-95-99.7) para una distribución normal, que establece que:

Aproximadamente el
68
%
68% de los datos se encuentra dentro de 1 desviación estándar del promedio.
Aproximadamente el
95
%
95% de los datos se encuentra dentro de 2 desviaciones estándar del promedio.
Aproximadamente el
99.7
%
99.7% de los datos se encuentra dentro de 3 desviaciones estándar del promedio.
Dado que nuestros porcentajes son superiores a estos valores, podemos inferir que la distribución de los tamaños de los meteoritos es menos dispersa alrededor de la media de lo que lo sería una distribución perfectamente normal. Es decir, la mayoría de los meteoritos tienen tamaños que se agrupan más estrechamente alrededor de la media en comparación con una distribución normal.